### PR TITLE
Issue 10285 - Enum grammar documentation is incorrect

### PR DESCRIPTION
**Pull request by Hackerpilot**
https://issues.dlang.org/show_bug.cgi?id=10285

This cleans up the enum declaration grammar and makes it more closely match DMD's implementation. Some rules that consisted of a single token were removed, and a new AnonymousEnumDeclaration rule was added to ensure that the "type identifer = assignExpression" syntax cannot appear in non-anonymous or typed enums.

### CODE REVIEW

File: enum.dd [hunk 1]
```diff
@@ -4,44 +4,45 @@ $(SPEC_S Enums,
...
-
-$(GNAME EnumTag):
-    $(I Identifier)
+    $(D enum) $(I Identifier) $(GLINK EnumBody)
+    $(D enum) $(I Identifier) $(D :) $(GLINK EnumBaseType) $(GLINK EnumBody)
+    $(AnonymousEnumDeclaration)
 
 $(GNAME EnumBaseType):
-    $(GLINK2 declaration, Type)
+    $(GLINK Type)
```
Comment (9rnsr) [inline enum.dd]: This change will disconnect the link.

File: enum.dd [hunk 2]
```diff
@@ -4,44 +4,45 @@ $(SPEC_S Enums,
...
 
 $(GNAME EnumBody):
-    $(GLINK EmptyEnumBody)
-    $(GLINK EnumMembersBody)
-
-$(GNAME EmptyEnumBody):
```
Comment (9rnsr) [inline enum.dd]: The `EmptyEnumBody` is used in later documentation. If you remove it, you should also fix the document text.

File: enum.dd [hunk 3]
```diff
@@ -4,44 +4,45 @@ $(SPEC_S Enums,
...
-    $(D ;)
-
-$(GNAME EnumMembersBody):
     $(D {) $(GLINK EnumMembers) $(D })
+    $(D ;)
 
 $(GNAME EnumMembers):
     $(GLINK EnumMember)
     $(GLINK EnumMember) $(D ,)
-    $(GLINK EnumMember) $(D ,) $(I EnumMembers)
```
Comment (9rnsr) [inline enum.dd]: Why is the repetition removed?

File: enum.dd [hunk 4]
```diff
@@ -4,44 +4,45 @@ $(SPEC_S Enums,
...
+    $(D enum) $(D {) $(GLINK EnumMembers) $(D })
+    $(D enum) $(D {) $(GLINK AnonymousEnumMembers) $(D })
+
+$(GNAME AnonymousEnumMembers):
+    $(GLINK AnonymousEnumMember)
+    $(GLINK AnonymousEnumMembers) $(D ,) $(GLINK AnonymousEnumMember)
+
+$(GNAME AnonymousEnumMember):
+    $(GLINK EnumMember)
+    $(GLINK Type) $(I Identifier) $(D =) $(ASSIGNEXPRESSION)
```
Comment (9rnsr) [inline enum.dd]: `Type` should use `$(GLINK2 declaration, Type) `.

File: enum.dd [hunk 5]
```diff
@@ -4,44 +4,45 @@ $(SPEC_S Enums,
...
-    $(GLINK2 declaration, Type) $(I Identifier) $(D =) $(ASSIGNEXPRESSION)
+
+$(GNAME AnonymousEnumDeclaration):
+    $(D enum) $(D :) $(GLINK EnumBaseType) $(D {) $(GLINK EnumMembers) $(D })
+    $(D enum) $(D {) $(GLINK EnumMembers) $(D })
+    $(D enum) $(D {) $(GLINK AnonymousEnumMembers) $(D })
+
+$(GNAME AnonymousEnumMembers):
+    $(GLINK AnonymousEnumMember)
+    $(GLINK AnonymousEnumMembers) $(D ,) $(GLINK AnonymousEnumMember)
```
Comment (9rnsr) [inline enum.dd]: Correction:
`$(GLINK AnonymousEnumMember) $(D ,) $(GLINK AnonymousEnumMembers)`

**Comment by 9rnsr** [comment]
Additional advice: If you hope to create a detailed Pull Request, you can change the grammar step by step by using multiple commits. For example:

1st commit: Remove `EnumTag` rule.
2nd commit: Remove `EmptyEnumBody` rule
3rd commit: Remove `EnumMembersBody` body to separate cases for named/anonymous enum
4th commitL Add `AnonymousXXX` rules for issue 10285

The stepped commit log would be much easier to review the correctness of the PR.

File: enum.dd (hunk 3 above)
Comment (Hackerpilot) [inline enum.dd]: That was a mistake. It's back now.

File: enum.dd (hunk 2 above)
Comment (Hackerpilot) [inline enum.dd]: The documentation text has been updated.

File: enum.dd (hunk 1 above)
Comment (Hackerpilot) [inline enum.dd]: Fixed.

File: enum.dd (hunk 5 above)
Comment (Hackerpilot) [inline enum.dd]: Fixed.

**Comment by 9rnsr** [comment]
LGTM.

**Pull request merged**